[ci] Add `golang_deps_generate` step to main branch builds

sgnn7 opened the pull request:
### What does this PR do?

Runs `inv agent.build-dep-tree` on main or release branches and saves the artifacts

### Motivation

We want to preserve and automate Golang dependency generation artifacts
for the Agent so this change ensures that we do that for all stable and
main branch builds.

### Additional Notes

This is a rebase of a PR made much earlier but it required Golang 1.16 on the main
branch to work.

### Describe how to test your changes

Check that main/release branches on GitLab run `golang_deps_generate` step under `source tests` group and contain an archived `dependency_tree.tgz` file

### Checklist

- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

sgnn7 committed:
[ci] Add `golang_deps_generate` step to main branch builds

We want to preserve and automate golang dependency generation artifacts
for the Agenbt so this change ensures that we do that for all stable and
main branch builds.

mx-psi commented on .gitlab/source_test/golang_deps_generate.yml:
```diff
@@ -0,0 +1,22 @@
+    ARCH: "x64"
+  before_script:
+    - !reference [.retrieve_linux_go_deps]
+    - python3 -m pip install --upgrade --ignore-installed pip setuptools
```
We don't usually need to do `--ignored-installed`, why do we need it here?

sgnn7 replied:
Copy/paste from https://github.com/DataDog/datadog-agent/blob/main/.gitlab/source_test/linux.yml#L128. I don't know for sure if it's needed though.

mx-psi commented on .gitlab/source_test/golang_deps_generate.yml:
```diff
@@ -0,0 +1,22 @@
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64:$DATADOG_AGENT_BUILDIMAGES
+  tags: ["runner:main"]
+  needs: ["linux_x64_go_deps"]
+  variables:
+    ARCH: "x64"
```
```suggestion
```

mx-psi commented on .gitlab/source_test/golang_deps_generate.yml:
```diff
@@ -0,0 +1,22 @@
+  stage: source_test
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64:$DATADOG_AGENT_BUILDIMAGES
+  tags: ["runner:main"]
+  needs: ["linux_x64_go_deps"]
```
If this job really needs the `linux_x64_go_deps` output it should do something with it

mx-psi commented on .gitlab/source_test/golang_deps_generate.yml:
```diff
@@ -0,0 +1,22 @@
+  script:
+    - inv agent.build-dep-tree
+    - tar -czf dependency_tree.tgz dependency_tree.txt
+  artifacts:
```
We should set explicitly the `expires_in` property here. Do we want to keep them forever?

sgnn7 replied:
I think we actually may want to keep them around forever (or at least a very long time). I don't know if we can be sure that we won't need an idea about how an artifact was made (especially in regards to the fed sector requirements around SBOMs)

mx-psi commented on .gitlab/source_test/golang_deps_generate.yml:
```diff
@@ -0,0 +1,22 @@
+    - python3 -m pip install -r requirements.txt
+  script:
+    - inv agent.build-dep-tree
+    - tar -czf dependency_tree.tgz dependency_tree.txt
```
Do we really need to `tar` the output? We don't usually do this unless we have several files that we reuse later in the pipeline (not the case here IIUC)

sgnn7 replied:
The resulting file is a plain text >100MB file. It seemed wasteful not to compress the artifact as it results in a 3-4MB artifact instead.

mx-psi replied:
Fair enough :)

sgnn7 committed:
Update .gitlab/source_test/golang_deps_generate.yml

Co-authored-by: Pablo Baeyens <pablo.baeyens@datadoghq.com>

sgnn7 committed:
[ci] Make golang dependency artifact never expire

Since we will need this artifact for a long time, we don't want it to
expire and this makes that explicitly defined.

sgnn7 commented:
Failing appveyor test is unrelated to the PR. Merging.
